src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/Handlers/Connection/LoginHandler.cs

```diff
@@ -0,0 +1,636 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Buffers;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Data.Common;
+using Microsoft.Data.SqlClient;
+using Microsoft.Data.SqlClientX.Handlers.Connection.Login;
+using Microsoft.Data.SqlClientX.IO;
+
+namespace Microsoft.Data.SqlClientX.Handlers.Connection
+{
+    /// <summary>
+    /// A handler for Sql Server login. 
+    /// </summary>
+    internal class LoginHandler : ContextHandler<ConnectionHandlerContext>
+    {
+        // NIC address caching
+        private static byte[] s_nicAddress;             // cache the NIC address from the registry
+        private readonly GlobalTransactionsFeature _globalTransactionsFeature;
+        private readonly DataClassificationFeature _dataClassificationFeature;
+        private readonly Utf8Feature _utf8SupportFeature;
+        private readonly SqlDnsCachingFeature _sqlDnsCachingFeature;
+        private readonly SessionRecoveryFeature _sessionRecoveryFeature;
+        private readonly FedAuthFeature _fedAuthFeature;
+        private readonly TceFeature _tceFeature;
+
+        /// <summary>
+        /// Parameterless constructor for the login handler.
+        /// </summary>
+        public LoginHandler()
```

Comment:
Can LoginHandler be initialized with context instead of passing context in the handler?
I'm wondering if these features can be initialized with context instead?

Reply:
It doesn't fit into the context-less design of handlers, today the orchestrator assumes that the handlers are stateless, and states are passed during Handle. 

The drawback is to instantiate a new handler for every connection request, which is unnecessary.

Reply:
That seems fine then.

Reply:
In the current stateless design, there is anyhow going to be one only one login handler.

Reply:
The reason I want to stay away from the classic singletons is because it doesn't lend itself well to unit testing.